libraries/chain/dbs_atomicswap.cpp

```diff
@@ -0,0 +1,176 @@
+#include <scorum/chain/dbs_atomicswap.hpp>
+#include <scorum/chain/database.hpp>
+#include <scorum/chain/dbs_account.hpp>
+#include <scorum/chain/account_object.hpp>
+#include <scorum/protocol/atomicswap_helper.hpp>
+
+using namespace scorum::protocol;
+
+namespace scorum {
+namespace chain {
+
+dbs_atomicswap::dbs_atomicswap(database& db)
+    : _base_type(db)
+{
+}
+
+dbs_atomicswap::atomicswap_contracts_refs_type dbs_atomicswap::get_contracts(const account_object& owner) const
+{
+    atomicswap_contracts_refs_type ret;
+
+    const auto& idx
+        = db_impl().get_index<atomicswap_contract_index>().indices().get<by_owner_name>().equal_range(owner.name);
+    for (auto it = idx.first; it != idx.second; ++it)
+    {
+        ret.push_back(std::cref(*it));
+    }
+
+    return ret;
+}
+
+const atomicswap_contract_object&
+dbs_atomicswap::get_contract(const account_object& from, const account_object& to, const std::string& secret_hash) const
+{
+    FC_ASSERT(!secret_hash.empty(), "Empty secret hash.");
+
+    try
+    {
+        atomicswap::hash_index_type contract_hash = atomicswap::get_contract_hash(from.name, to.name, secret_hash);
+        return db_impl().get<atomicswap_contract_object, by_contract_hash>(contract_hash);
+    }
+    FC_CAPTURE_AND_RETHROW((from.name)(to.name)(secret_hash))
+}
+
+const atomicswap_contract_object& dbs_atomicswap::create_contract(atomicswap_contract_type tp,
+                                                                  const account_object& owner,
+                                                                  const account_object& recipient,
+                                                                  const asset& amount,
+                                                                  const std::string& secret_hash,
+                                                                  const optional<std::string>& metadata)
+{
+    FC_ASSERT(amount.symbol == SCORUM_SYMBOL, "Invalid asset type (symbol).");
```

Comment:
You putted in data service evaluator logic. All this checking stuff should be in evaluator. I believe it is better do not mix dataservices. Data service is a set of getters/setters for DB objects. In your implementation you blur business logic instead of keeping it in one place.